client/parser/test_helper.go

```diff
@@ -73,6 +74,23 @@ func executeParseTestCases(t *testing.T, testCases ...parseTestCase) {
 			}
 		}
 
+		// Test formatter by ensuring round-trip formatted + parsed AST is same
+		buffer := bytes.Buffer{}
+		err = FormatASTToCode(actualAST, &buffer)
+		if err != nil {
+			t.Errorf("%v\n", err)
+			return
+		}
+		formattedAST, err := Parse(
+			// The newly formatted file will have different source context's
+			testCase.label, buffer.String(), SuppressSourceContext)
+		if diff := deep.Equal(actualAST, formattedAST); diff != nil {
+			t.Error(testCase.label)
+			for _, diffItem := range diff {
+				t.Errorf("%v", diffItem)
+			}
+		}
```

Comment:
Nice idea, testing that the formatted code parses to an equivalent AST as the original code :+1:

Reply:
Thanks, I think this invariant should be maintained in all cases....

There is one interesting thing I realized which is when transforming `model.AttributeUpdate{model.PartUpdate}` back into Lisp for formatting, we have to arbitrarily pick one of possibly multiple Lisp defn's (e.x. "transposition" vs "transpose"). This doesn't affect testing the formatter though as we don't go that deep in this parser test case.

Reply:
Hmm, this seems like a hard problem, because the idea with the Lisp code is that it evaluates code into a ScoreUpdate. But we can't accurately turn a ScoreUpdate back into the code that produced it, because there can be many different code expressions that evaluate to the same ScoreUpdate. This thing about attribute aliases (e.g. "vol" vs. "volume") is a very simple example, but theoretically in the future, we could decide to add more built-in functions to the Lisp language, which could make it all the more difficult (impossible, actually) to generate the correct code from a ScoreUpdate.

Imagine, for example, that we added math operators, so that you could express `(vol 100)` in many ways:

```lisp
(vol (+ 50 50))
(vol (* 10 10))
(vol (- 200 100))
```

I think it would be problematic if the formatter simplified Lisp code expressions like the above as `(vol 100)`, or even `(volume 100)`, because the job of a formatter isn't to evaluate or simplify code, it's to format it.

I don't think we necessarily need to do anything about this right now (maybe a FIXME/TODO comment would suffice), but one idea of a potential solution is to add an optional Metadata property on all ScoreUpdates, and store the Lisp expression there after we evaluate the expression, so that we can retain it for the purposes of converting back to code. Actually, this Metadata property might help us in other contexts as well, whenever we have any kind of situation where multiple possible AST nodes could produced the same ScoreUpdate.

Reply:
Importantly, the Metadata would be ignored when doing deep diffs.

I got this metadata idea from Clojure. Clojure lets you attach metadata to values, and the metadata can be inspected as needed, but otherwise, the values work exactly the same way as they always do:

```clojure
user=> (def a ^{:my-property "my value"} {:x 42})
#'user/a
user=> (def b ^{:my-property "different value"} {:x 42})
#'user/b
user=> a
{:x 42}
user=> b
{:x 42}
user=> (meta a)
{:my-property "my value"}
user=> (meta b)
{:my-property "different value"}
;; a and b are considered equal, despite having different metadata
user=> (= a b)
true
```

Reply:
I'm a bit confused, but I'm thinking we're ok right now and it seems we don't have to do anything.
- For `alda format`, we never actually go past the `ASTNode` layer to `ScoreUpdate`s. This means the Lisp is never evaluated or simplified. So we should be good here. The point of AST I guess is to only be the structure of the language.
- For `alda import`, we generate whatever ScoreUpdates we want. Right now, every `ScoreUpdate` can be generated to `ASTNode`. As long as we cover this single use case of code generation, we're good for now.
- Down the line if we run into issues, we can make changes as necessary.

What TODO exactly are you trying to point out? I can see how Lisp eval can be "lossy" of the original code, but that's not a problem with formatting because we only go to AST. Perhaps there can be issues down the line generating code from complicated lisp `ScoreUpdate`s, but surely we can just generate something reasonable.

Sorry if I'm misunderstanding, this Lisp part has always confused me.

Reply:
Sorry David, I think I confused both of us! I'm thinking about it more now, and I think you're right that this is not an issue. It doesn't look like we would ever need to go from AST -> ScoreUpdates -> AST -> formatted code. Disregard my rambling above :smile: